### PR TITLE
Bug fix/fixes asan uninitialized by line

### PR DESCRIPTION
**Pull request by mchacki**
Fixes use after free
and potentially uninitialized variables.

http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/3764/

### CODE REVIEW

File: arangod/Aql/ModificationExecutorTraits.cpp [hunk 1]
```diff
@@ -658,10 +658,9 @@ bool UpdateReplace<ModType>::doModifications(ModificationExecutorInfos& info,
   std::string key;
   std::string rev;
   auto* trx = info._trx;
-  const RegisterId& inDocReg = info._input1RegisterId.value();
-  const bool hasKeyVariable = info._input2RegisterId.has_value();
-  const RegisterId& keyReg = hasKeyVariable ? info._input2RegisterId.value()
-                                            : 42 /*invalid but unused*/;
```
Comment (jsteemann) [inline arangod/Aql/ModificationExecutorTraits.cpp]: wtf

File: arangod/Aql/SingleRemoteModificationExecutor.cpp [hunk 2]
```diff
@@ -109,24 +108,19 @@ bool SingleRemoteModificationExecutor<Modifier>::doSingleRemoteModificationOpera
...
 
   VPackBuilder inBuilder;
   VPackSlice inSlice = VPackSlice::emptyObjectSlice();
-  if (_info._input1RegisterId.has_value()) {  // IF NOT REMOVE OR SELECT
-    // if (_buffer.size() < 1) {
-    //  THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_ARANGO_DOCUMENT_NOT_FOUND,
-    //                                 "missing document reference in Register");
-    //}
-    AqlValue const& inDocument = input.getValue(_info._input1RegisterId.value());
+  if (_info._input1RegisterId == ExecutionNode::MaxRegisterId) {  // IF NOT REMOVE OR SELECT
```
Comment (jsteemann) [inline arangod/Aql/SingleRemoteModificationExecutor.cpp]: doesn't this `if` revert the old behavior?
```suggestion
  if (_info._input1RegisterId != ExecutionNode::MaxRegisterId) {  // IF NOT REMOVE OR SELECT
```

**Comment by ObiWahn** [comment]
I like the change a lot! The extra flag was a nuisance. What I do not understand is the equality check on `MaxRegisterId` that is defined to be 1000 not the maximum of the interger-type. Therefore I would assume that the check should be `< MaxId` or `>=  MaxId` if the value is used as register. Or is there a way for integers above `MaxRegisterId` to be used as valid register-ids? Because then the variable's name would be misleading.

**Comment by mchacki** [comment]
@ObiWahn no MaxRegisterId is the first invalid register.
Right now this is coded to 1000 Registers you can use add a time.
But that is why we should rely on this number s.t. we can modify one constant if we want to enlarge/reduce the amount of possible registers.

**Pull request merged**